File: src/codes/ref-synergy/ts/proxy-strategy.ts
interface NotificationStrategy {
  send(user: string, message: string): void;
}

class EmailStrategy implements NotificationStrategy {
  send(user: string, message: string): void {
    console.log(`[Email] To: ${user}, Message: ${message}`);
  }
}

class SlackStrategy implements NotificationStrategy {
  send(user: string, message: string): void {
    console.log(`[Slack] To: ${user}, Message: ${message}`);
  }
}

interface Notifier {
  notify(user: string, message: string): void;
}

class RealNotifier implements Notifier {
  constructor(private strategy: NotificationStrategy) {}

  notify(user: string, message: string): void {
    this.strategy.send(user, message);
  }
}

class NotifierProxy implements Notifier {
  constructor(private real: Notifier) {}

  notify(user: string, message: string): void {
    if (!user.includes("@")) {
      console.log("Invalid user. Notification aborted.");
      return;
    }
    console.log("Proxy: Access check passed.");
    this.real.notify(user, message);
    console.log("Proxy: Notification sent.");
  }
}

// Usage
const realNotifier = new RealNotifier(new EmailStrategy());
const proxy = new NotifierProxy(realNotifier);
proxy.notify("taro@example.com", "Disk space warning");

File: src/codes/ref-synergy/py/proxy-strategy.py
from abc import ABC, abstractmethod

class NotificationStrategy(ABC):
    @abstractmethod
    def send(self, user: str, message: str) -> None:
        pass

class EmailStrategy(NotificationStrategy):
    def send(self, user: str, message: str) -> None:
        print(f"[Email] To: {user}, Message: {message}")

class SlackStrategy(NotificationStrategy):
    def send(self, user: str, message: str) -> None:
        print(f"[Slack] To: {user}, Message: {message}")

class Notifier(ABC):
    @abstractmethod
    def notify(self, user: str, message: str) -> None:
        pass

class RealNotifier(Notifier):
    def __init__(self, strategy: NotificationStrategy):
        self.strategy = strategy

    def notify(self, user: str, message: str) -> None:
        self.strategy.send(user, message)

class NotifierProxy(Notifier):
    def __init__(self, real: Notifier):
        self.real = real

    def notify(self, user: str, message: str) -> None:
        if "@" not in user:
            print("Invalid user. Notification aborted.")
            return
        print("Proxy: Access check passed.")
        self.real.notify(user, message)
        print("Proxy: Notification sent.")

# Usage
real_notifier = RealNotifier(EmailStrategy())
proxy = NotifierProxy(real_notifier)
proxy.notify("taro@example.com", "Disk space warning")
